components/forms/src/api.rs

```diff
@@ -0,0 +1,248 @@
+    // Invalid and negative timestamps are all replaced with 0. Eventually we
+    // should investigate replacing values that are unreasonable but still fit
+    // in an i64 (a date 1000 years in the future, for example), but
+    // appropriately handling that is complex.
+    Ok(i64::deserialize(deserializer).unwrap_or_default().max(0))
+}
+
+/*
+    TODO:
+        - get all
+        - update
+        - delete
+        - check if deleted on create and/or update?
+*/
+
+fn add_address(tx: &Transaction<'_>, mut address: Address) -> Result<Address> {
+    let d = time::SystemTime::now().duration_since(time::UNIX_EPOCH).unwrap_or_default();
+    let now_ms = (d.as_secs() as i64) * 1000 + (i64::from(d.subsec_nanos()) / 1_000_000);
+
+    address.guid = Guid::random();
+    address.time_created = now_ms;
+    address.time_last_used = 0;
+    address.time_last_modified = now_ms;
+    address.times_used = 0;
+
+    tx.execute_named_cached(
+        "INSERT OR IGNORE INTO addresses_data (
+            guid,
+            given_name,
+            additional_name,
+            family_name,
+            organization,
+            street_address,
+            address_level3,
+            address_level2,
+            address_level1,
+            postal_code,
+            country,
+            tel,
+            email,
+            time_created,
+            time_last_used,
+            time_last_modified,
+            times_used,
+            sync_change_counter
+        ) VALUES (
+            :guid,
+            :given_name,
+            :additional_name,
+            :family_name,
+            :organization,
+            :street_address,
+            :address_level3,
+            :address_level2,
+            :address_level1,
+            :postal_code,
+            :country,
+            :tel,
+            :email,
+            :time_created,
+            :time_last_used,
+            :time_last_modified,
+            :times_used,
+            1
+        )",
+        rusqlite::named_params! {
+            ":guid": address.guid,
+            ":given_name": address.given_name,
+            ":additional_name": address.additional_name,
+            ":family_name": address.family_name,
+            ":organization": address.organization,
+            ":street_address": address.street_address,
+            ":address_level3": address.address_level3,
+            ":address_level2": address.address_level2,
+            ":address_level1": address.address_level1,
+            ":postal_code": address.postal_code,
+            ":country": address.country,
+            ":tel": address.tel,
+            ":email": address.email,
+            ":time_created": address.time_created,
+            ":time_last_used": address.time_last_used,
+            ":time_last_modified": address.time_last_modified,
+            ":times_used": address.times_used,
+        },
+    )?;
+    Ok(address)
+}
+
+fn get_address(conn: &Connection, guid: &Guid) -> Result<Option<Address>> {
```

Comment:
Let’s return an `AddressDoesNotExist` error if the address doesn’t exist, and simplify the return signature to just `Result<Address>`. I’m thinking, if the caller wants to distinguish between address not found and other kinds of DB errors, it can destructure the error instead of doing a double-unwrap.

Reply:
This was an oversight on my part but the [query_row](https://docs.rs/rusqlite/0.24.0/rusqlite/struct.Connection.html#method.query_row) function I'm using to execute my sql statement actually returns an error if no rows are found. I'm going to remove the `Option` from the return type but let me know if I should convert the `QueryReturnedNoRows` type to a custom error.